phoenix-core-client/src/main/java/org/apache/phoenix/query/ConnectionQueryServicesImpl.java

```diff
@@ -6248,4 +6320,170 @@ public synchronized PhoenixTransactionClient initTransactionClient(Provider prov
     public List<LinkedBlockingQueue<WeakReference<PhoenixConnection>>> getCachedConnections() {
       return connectionQueues;
     }
+
+    /**
+     * Invalidate metadata cache from all region servers for the given list of
+     * InvalidateServerMetadataCacheRequest.
+     * @throws Throwable
+     */
+    public void invalidateServerMetadataCache(List<InvalidateServerMetadataCacheRequest> requests)
+            throws Throwable {
+        boolean invalidateCacheEnabled =
+                config.getBoolean(PHOENIX_METADATA_INVALIDATE_CACHE_ENABLED,
+                        DEFAULT_PHOENIX_METADATA_INVALIDATE_CACHE_ENABLED);
+        if (!invalidateCacheEnabled) {
+            LOGGER.info("Skip invalidating server metadata cache since conf property"
+                    + " phoenix.metadata.invalidate.cache.enabled is set to false");
+            return;
+        }
+        if (!QueryUtil.isServerConnection(props)) {
+            LOGGER.warn(INVALIDATE_SERVER_METADATA_CACHE_EX_MESSAGE);
+            throw new Exception(INVALIDATE_SERVER_METADATA_CACHE_EX_MESSAGE);
+        }
+
+        metricsMetadataCachingSource.incrementMetadataCacheInvalidationOperationsCount();
+        Admin admin = getInvalidateMetadataCacheConnection().getAdmin();
+        // This will incur an extra RPC to the master. This RPC is required since we want to
+        // get current list of regionservers.
+        Collection<ServerName> serverNames = admin.getRegionServers(true);
+        PhoenixStopWatch stopWatch = new PhoenixStopWatch().start();
+        try {
+            invalidateServerMetadataCacheWithRetries(admin, serverNames, requests, false);
+            metricsMetadataCachingSource.incrementMetadataCacheInvalidationSuccessCount();
+        } catch (Throwable t) {
+            metricsMetadataCachingSource.incrementMetadataCacheInvalidationFailureCount();
+            throw t;
+        } finally {
+            metricsMetadataCachingSource
+                    .addMetadataCacheInvalidationTotalTime(stopWatch.stop().elapsedMillis());
+        }
+    }
+
+    /**
+     * Invalidate metadata cache on all regionservers with retries for the given list of
+     * InvalidateServerMetadataCacheRequest. Each InvalidateServerMetadataCacheRequest contains
+     * tenantID, schema name and table name.
+     * We retry once before failing the operation.
+     *
+     * @param admin
+     * @param serverNames
+     * @param invalidateCacheRequests
+     * @param isRetry
+     * @throws Throwable
+     */
+    private void invalidateServerMetadataCacheWithRetries(Admin admin,
+            Collection<ServerName> serverNames,
+            List<InvalidateServerMetadataCacheRequest> invalidateCacheRequests,
+            boolean isRetry) throws Throwable {
+        RegionServerEndpointProtos.InvalidateServerMetadataCacheRequest protoRequest =
+                getRequest(invalidateCacheRequests);
+        // TODO Do I need my own executor or can I re-use QueryServices#Executor
```

Comment:
This came up in discussion today and the recommendation is to create a separate thread pool for executing cache invalidations. I can create a subtask for the same and add the changes to this PR.
@shahrs87 @tkhurana @kadirozde

Reply:
Assuming that these new methods are only called at the server side, I think we should move this code to a separate or existing server side class.

Reply:
Since the cache invalidation code is in CQSI and we would want to avoid creating a new thread pool for every DDL operation, how do you suggest we save the executor? We can have an `executor` in CQSI but we would not want client side CQSI to create an executor unnecessarily - or is that okay?

Reply:
@kadirozde This code did start out in a separate class but I remember @shahrs87 and @jpisaac had some reasons to move it to CQSI. I could not find which PR that was. @shahrs87 Can you recall those details?

Reply:
This was the PR - https://github.com/apache/phoenix/pull/1748
@kadirozde Please see comments here https://github.com/apache/phoenix/pull/1748#discussion_r1419676088